### PR TITLE
Add Apache Thrift detection

### PR DESCRIPTION
**Pull request by 0xA50C1A1**
[Apache Thrift](https://github.com/apache/thrift) is a commonly used RPC framework.

This detector is quite simple and only checks the HTTP content-type field or looks for the Thrift strict or compact header in TCP or UDP packets. This can give false-positive results in theory, but I guess such accuracy is enough in practice.

### CODE REVIEW

**Comment by IvanNardi** [comment]
@0xA50C1A1 , I didn't look at your code yet, but could you attached a trace with also an example of detection over HTTP, please?
BTW: you need to add the new .c file to the VisualStudio configuration file....

**Comment by 0xA50C1A1** [comment]
> @0xA50C1A1 , I didn't look at your code yet, but could you attached a trace with also an example of detection over HTTP, please? BTW: you need to add the new .c file to the VisualStudio configuration file....

I don't have a PCAP sample with Thrift-over-HTTP traffic yet, but I'm looking for.

File: src/lib/protocols/thrift.c [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+static void ndpi_int_thrift_add_connection(struct ndpi_detection_module_struct *ndpi_struct, 
+					   struct ndpi_flow_struct *flow)
+{
+    ndpi_set_detected_protocol(ndpi_struct, flow, NDPI_PROTOCOL_THRIFT, NDPI_PROTOCOL_UNKNOWN, NDPI_CONFIDENCE_DPI);
+}
+
+static void ndpi_search_thrift_tcp_udp(struct ndpi_detection_module_struct *ndpi_struct, struct ndpi_flow_struct *flow)
+{
+    struct ndpi_packet_struct *packet = &ndpi_struct->packet;
+
```
Comment (IvanNardi) [inline src/lib/protocols/thrift.c:90]: Could you try, at least, something like that:
```
/* As a HTTP sub-protocol */
if(flow->detected_protocol_stack[0] == NDPI_PROTOCOL_HTTP ||
   flow->detected_protocol_stack[1] == NDPI_PROTOCOL_HTTP) {
  /* Content line stuff */
} else  { /* Standard dissector */
  if (packet->payload_packet_len > 5) {

  }
}
NDPI_EXCLUDE_PROTO(ndpi_struct, flow);
```

**Review by IvanNardi — CHANGES_REQUESTED** [review]
Even ignoring the HTTP change, it seems there a lots of false positives....

**Comment by 0xA50C1A1** [comment]
Sorry, I was too hasty with a pull request, so I'll change it to a draft until I fix this dissector.

File: src/include/ndpi_protocol_ids.h [hunk 2]
```diff
@@ -361,6 +361,7 @@ typedef enum {
   NDPI_PROTOCOL_ELASTICSEARCH         = 330,
   NDPI_PROTOCOL_TUYA_LP               = 331, /* TUYA LAN Protocol; IoT OS: https://github.com/tuya/tuya-iotos-embeded-sdk-wifi-ble-bk7231n */
   NDPI_PROTOCOL_TPLINK_SHP            = 332, /* TP-LINK Smart Home Protocol */
+  NDPI_PROTOCOL_THRIFT                = 333, /* Apache Thrift */
```
Comment (utoni) [inline src/include/ndpi_protocol_ids.h:364]: Isn't there a `*_FREE_*` protocol you can use? (Sorry, can not verify at the moment as I have just my mobile available.)

Comment (IvanNardi) [reply src/include/ndpi_protocol_ids.h:364]: +1 on using a `*_FREE_*` id

**Comment by utoni** [comment]
@0xA50C1A1 please rebase to the current dev branch

**Comment by utoni** [comment]
Superseeded by #2007.

**Pull request closed without merging**